From f216a6e787f89b7545e28494f624e0ce997900ea Mon Sep 17 00:00:00 2001
From: Janne Grunau <j@jannau.net>
Date: Sat, 3 Dec 2022 22:19:44 +0100
Subject: [PATCH] arm64: dts: apple: t8112: Add dp-altmode dts hacks

Enable DP alt mode for the front left USB-C port of Macbook Air 13 (M2,
13/15-inch) and Macbook Pro (M2, 13-inch).
Using dcp as display coproc since dcpext is used for the HDMI port.

Signed-off-by: Janne Grunau <j@jannau.net>
---
 arch/arm64/boot/dts/apple/t8112-j413.dts  |  2 +
 arch/arm64/boot/dts/apple/t8112-j415.dts  |  2 +
 arch/arm64/boot/dts/apple/t8112-j473.dts  | 47 +++++++++++++++++---
 arch/arm64/boot/dts/apple/t8112-j493.dts  |  2 +
 arch/arm64/boot/dts/apple/t8112-jxxx.dtsi | 52 +++++++++++++++++++++++
 5 files changed, 100 insertions(+), 5 deletions(-)

diff --git a/arch/arm64/boot/dts/apple/t8112-j413.dts b/arch/arm64/boot/dts/apple/t8112-j413.dts
index 5ae973bc79fef4..82f54d603167f5 100644
--- a/arch/arm64/boot/dts/apple/t8112-j413.dts
+++ b/arch/arm64/boot/dts/apple/t8112-j413.dts
@@ -9,6 +9,8 @@
 
 /dts-v1/;
 
+#define ENABLE_DCPEXT_TYPEC
+
 #include "t8112.dtsi"
 #include "t8112-jxxx.dtsi"
 #include <dt-bindings/leds/common.h>
diff --git a/arch/arm64/boot/dts/apple/t8112-j415.dts b/arch/arm64/boot/dts/apple/t8112-j415.dts
index 20a64ecc122d87..c9f0d1a879452e 100644
--- a/arch/arm64/boot/dts/apple/t8112-j415.dts
+++ b/arch/arm64/boot/dts/apple/t8112-j415.dts
@@ -9,6 +9,8 @@
 
 /dts-v1/;
 
+#define ENABLE_DCPEXT_TYPEC
+
 #include "t8112.dtsi"
 #include "t8112-jxxx.dtsi"
 #include <dt-bindings/leds/common.h>
diff --git a/arch/arm64/boot/dts/apple/t8112-j473.dts b/arch/arm64/boot/dts/apple/t8112-j473.dts
index 4fc96779806ea3..cb94c7faa03263 100644
--- a/arch/arm64/boot/dts/apple/t8112-j473.dts
+++ b/arch/arm64/boot/dts/apple/t8112-j473.dts
@@ -9,6 +9,8 @@
 
 /dts-v1/;
 
+#define ENABLE_DCPEXT_TYPEC
+
 #include "t8112.dtsi"
 #include "t8112-jxxx.dtsi"
 
@@ -22,7 +24,6 @@
 		/delete-property/ dcp;
 		dcpext = &dcpext;
 		ethernet0 = &ethernet0;
-		sio = &sio;
 		wifi0 = &wifi0;
 	};
 };
@@ -66,15 +67,51 @@
 	apple,dptx-phy = <5>;
 };
 
-/* remove once m1n1 enables sio nodes after setup */
-&sio {
-        status = "okay";
+#ifdef ENABLE_DCPEXT_TYPEC
+/ {
+	aliases {
+		dcp = &dcp;
+	};
+};
+
+&display {
+	/* dispext0_dart needs to be first for m1n1's */
+	iommus = <&dispext0_dart 0>, <&disp0_dart 0>;
+};
+
+&disp0_dart {
+	status = "okay";
+};
+
+&dcp {
+	status = "okay";
+	apple,connector-type = "DP";
+
+	/* hacks */
+	apple,dptx-phy = <1>;
+	phys = <&atcphy1 PHY_TYPE_DP>;
+	phy-names = "dp-phy";
+	mux-controls = <&atcphy1_xbar 0>;
+	mux-control-names = "dp-xbar";
+	mux-index = <0>;
 };
 
-&dpaudio1 {
+&dcpext {
+	/delete-property/ mux-controls;
+	/delete-property/ mux-control-names;
+	/delete-property/ mux-index;
+};
+
+&dpaudio0 {
 	status = "okay";
 };
 
+&typec1 {
+	/* hacks */
+	displayport = <&dcp>;
+};
+#endif
+
 /*
  * Keep the power-domains used for the HDMI port on.
  */
diff --git a/arch/arm64/boot/dts/apple/t8112-j493.dts b/arch/arm64/boot/dts/apple/t8112-j493.dts
index fd3953fbb2fa0b..ce64fd1f77f881 100644
--- a/arch/arm64/boot/dts/apple/t8112-j493.dts
+++ b/arch/arm64/boot/dts/apple/t8112-j493.dts
@@ -9,6 +9,8 @@
 
 /dts-v1/;
 
+#define ENABLE_DCPEXT_TYPEC
+
 #include "t8112.dtsi"
 #include "t8112-jxxx.dtsi"
 #include <dt-bindings/leds/common.h>
diff --git a/arch/arm64/boot/dts/apple/t8112-jxxx.dtsi b/arch/arm64/boot/dts/apple/t8112-jxxx.dtsi
index fb957f785d82c5..a2631d28737124 100644
--- a/arch/arm64/boot/dts/apple/t8112-jxxx.dtsi
+++ b/arch/arm64/boot/dts/apple/t8112-jxxx.dtsi
@@ -18,6 +18,7 @@
 		disp0_piodma = &disp0_piodma;
 		serial0 = &serial0;
 		serial2 = &serial2;
+		sio = &sio;
 	};
 
 	chosen {
@@ -96,6 +97,11 @@
 			power-role = "dual";
 			data-role = "dual";
 
+#ifdef ENABLE_DCPEXT_TYPEC
+			/* hacks */
+			displayport = <&dcpext>;
+#endif
+
 			ports {
 				#address-cells = <1>;
 				#size-cells = <0>;
@@ -116,6 +122,52 @@
 	};
 };
 
+#ifdef ENABLE_DCPEXT_TYPEC
+/ {
+	aliases {
+		dcpext = &dcpext;
+	};
+};
+
+&display {
+	iommus = <&disp0_dart 0>, <&dispext0_dart 0>;
+};
+
+&dispext0_dart {
+	status = "okay";
+};
+
+&dcpext_dart {
+	status = "okay";
+};
+
+&dcpext_mbox {
+	status = "okay";
+};
+
+&dcpext {
+	status = "okay";
+	apple,connector-type = "DP";
+
+	/* hacks */
+	apple,dptx-phy = <1>;
+	phys = <&atcphy1 PHY_TYPE_DP>;
+	phy-names = "dp-phy";
+	mux-controls = <&atcphy1_xbar 0>;
+	mux-control-names = "dp-xbar";
+	mux-index = <2>;
+};
+
+/* remove once m1n1 enables sio nodes after setup */
+&sio {
+        status = "okay";
+};
+
+&dpaudio1 {
+	status = "okay";
+};
+#endif
+
 /* USB controllers */
 &dwc3_0 {
 	ports {
